flink-connector-mongodb/src/main/java/org/apache/flink/connector/mongodb/sink/config/MongoWriteOptions.java

```diff
@@ -0,0 +1,233 @@
+            checkArgument(
+                    numMaxActions == -1 || numMaxActions > 0,
+                    "Max number of buffered actions must be larger than 0.");
+            this.bulkFlushMaxActions = numMaxActions;
+            return this;
+        }
+
+        /**
+         * Sets the bulk flush interval, in milliseconds. You can pass -1 to disable it.
+         *
+         * @param intervalMillis the bulk flush interval, in milliseconds.
+         * @return this builder
+         */
+        public MongoWriteOptionsBuilder setBulkFlushIntervalMs(long intervalMillis) {
+            checkArgument(
+                    intervalMillis == -1 || intervalMillis >= 0,
+                    "Interval (in milliseconds) between each flush must be larger than "
+                            + "or equal to 0.");
+            this.bulkFlushIntervalMs = intervalMillis;
+            return this;
+        }
+
+        /**
+         * Sets the max retry times if writing records failed.
+         *
+         * @param maxRetryTimes the max retry times.
+         * @return this builder
+         */
+        public MongoWriteOptionsBuilder setMaxRetryTimes(int maxRetryTimes) {
+            checkArgument(
+                    maxRetryTimes >= 0,
+                    "The sink max retry times must be larger than or equal to 0.");
+            this.maxRetryTimes = maxRetryTimes;
+            return this;
+        }
+
+        /**
+         * Sets the retry interval if writing records to database failed.
+         *
+         * @param retryIntervalMs the retry time interval, in milliseconds.
+         * @return this builder
+         */
+        public MongoWriteOptionsBuilder setRetryInterval(long retryIntervalMs) {
+            checkArgument(
+                    retryIntervalMs > 0,
+                    "The retry interval (in milliseconds) must be larger than 0.");
+            this.retryIntervalMs = retryIntervalMs;
+            return this;
+        }
+
+        /**
+         * Sets the wanted {@link DeliveryGuarantee}. The default delivery guarantee is {@link
+         * DeliveryGuarantee#AT_LEAST_ONCE}
+         *
+         * @param deliveryGuarantee which describes the record emission behaviour
+         * @return this builder
+         */
+        public MongoWriteOptionsBuilder setDeliveryGuarantee(DeliveryGuarantee deliveryGuarantee) {
+            checkArgument(
+                    deliveryGuarantee != DeliveryGuarantee.EXACTLY_ONCE,
+                    "Mongo sink does not support the EXACTLY_ONCE guarantee.");
+            this.deliveryGuarantee = checkNotNull(deliveryGuarantee);
+            return this;
+        }
+
+        /**
+         * Sets the parallelism of the Mongo sink operator. By default, the parallelism is
+         * determined by the framework using the same parallelism of the upstream chained operator.
+         */
+        public MongoWriteOptionsBuilder setParallelism(int parallelism) {
```

Comment:
@zentol 

The parallelism parameter refers to jdbc and hbase connector.
In fact, I am also a little confused about this.
Should we delete the `parallelism` parameter ?

```java
/** Options for the JDBC connector. */
@Internal
public class JdbcConnectorOptions extends JdbcConnectionOptions {
    private final String tableName;
    private final JdbcDialect dialect;
    private final @Nullable Integer parallelism;
}

public class JdbcDynamicTableSink {
    @Override
    public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
        final TypeInformation<RowData> rowDataTypeInformation =
                context.createTypeInformation(physicalRowDataType);
        final JdbcOutputFormatBuilder builder = new JdbcOutputFormatBuilder();

        builder.setJdbcOptions(jdbcOptions);
        builder.setJdbcDmlOptions(dmlOptions);
        builder.setJdbcExecutionOptions(executionOptions);
        builder.setRowDataTypeInfo(rowDataTypeInformation);
        builder.setFieldDataTypes(
                DataType.getFieldDataTypes(physicalRowDataType).toArray(new DataType[0]));
        return SinkFunctionProvider.of(
                new GenericJdbcSinkFunction<>(builder.build()), jdbcOptions.getParallelism());
    }
}
```

```java
/** Options for HBase writing. */
@Internal
public class HBaseWriteOptions implements Serializable {
    private final long bufferFlushMaxSizeInBytes;
    private final long bufferFlushMaxRows;
    private final long bufferFlushIntervalMillis;
    private final Integer parallelism;
}

public class HBaseDynamicTableSink {
    @Override
    public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
        HBaseSinkFunction<RowData> sinkFunction =
                new HBaseSinkFunction<>(
                        tableName,
                        hbaseConf,
                        new RowDataToMutationConverter(hbaseTableSchema, nullStringLiteral),
                        writeOptions.getBufferFlushMaxSizeInBytes(),
                        writeOptions.getBufferFlushMaxRows(),
                        writeOptions.getBufferFlushIntervalMillis());
        return SinkFunctionProvider.of(sinkFunction, writeOptions.getParallelism());
    }
}
```

Reply:
I'll ask around why this parallelism setting exists. :+1:

Reply:
So it seems that `FactoryUtil#SINK_PARALLELISM` is a common constant that factories can use, and the factories just pass this parallelism to the `*TableSink` which then _again_ forwards it to `SinkFunctionProvider`, from which the Table API finally derives the parallelism.

Seems to be a bit of a mess in my eyes.
I'd suggest to not put it into the `WriteOptions` (because that's shared across APIs) and keep it focused on things that the _actual_ sink needs.
We can pass the parallelism explicitly to the TableSink constructor instead.

Reply:
Got it.  
I'll pass the parallelism explicitly to the TableSink constructor instead.

Reply:
Fixed.